xc/domain.ml

```diff
@@ -980,39 +1002,97 @@ module Suspend_restore_emu_manager : SUSPEND_RESTORE = struct
+                wrap (fun () -> parse_result response) >>= fun result ->
+                let emu = emu_of_result result in
+                (* Wake up the reader that has requested a result for this emu *)
+                if List.mem_assoc emu !thread_requests then begin
+                  let wakeup = List.assoc emu !thread_requests in
+                  Event.send wakeup () |> Event.sync;
+                  loop (result :: results)
+                end else begin
+                  error "Received unexpected response from emu-manager";
+                  `Error Domain_restore_failed
+                end
+              with End_of_file ->
+                debug "Finished emu-manager result processing";
+                return results
+            in
+            loop [] >>= fun results ->
+            (* We are only really interested in the result from xenguest *)
+            List.fold_left (function acc -> function
+                | Xenguest_result (store, console) -> return (Some (store, console))
+                | _ -> acc
+              ) (return None) results
+          in
+          let start_reader_thread fd =
+            (* Start a reader thread on the given fd. Add a channel back to the
+               				 * main thread for status reporting *)
+            debug "Starting reader thread (fd=%d)" (Obj.magic fd);
+            let ch = Event.new_channel () in
+            let th = Thread.create (fun () ->
+                wrap_exn (fun () -> process_header fd (return ()))
+                |> Event.send ch
+                |> Event.sync
+              ) () in
+            th, ch
+          in
+          let receive_thread_status threads_and_channels =
+            (* Receive the status from all reader threads and let them exit *)
+            fold (fun (th, ch) _ ->
+                let status = Event.receive ch |> Event.sync in
+                Thread.join th;
+                status
+              ) threads_and_channels () >>= fun () ->
+            debug "Reader threads completed successfully";
+            return ()
           in
           let res =
-            match vgpu_fd with
-            | Some fd when fd <> main_fd ->
-              debug "Starting vGPU resume thread";
-              let th = Thread.create (process_header fd) (return None) in
-              let res = process_header main_fd (return None) in
-              Thread.join th;
-              debug "vGPU resume thread finished";
-              res
-            | _ ->
-              process_header main_fd (return None)
+            (* Start a reader thread on each fd *)
+            let threads_and_channels = List.map start_reader_thread fds in
+            (* Handle results returned by emu-manager *)
+            handle_results () >>= fun result ->
+            (* Wait for reader threads to complete *)
+            receive_thread_status threads_and_channels >>= fun () ->
+            (* And we are done! *)
+            return result
```

Comment:
There is a potential issue here due to the fact that bind short-circuits when an error happens.
In case of error in the `handle_results` phase, we are going to skip the syncronization and completion of the reader threads and we can potentially start leaking threads.

A possible fix could be to move the `let threads_and_channels = ... in` outsude the `let res = ...` block and use the `threads_and_chanels` in the `Error e` match case below for cleanup operations. With @robhoes we mentioned either sending the necessary signals by hand to make the thread fail or try and check the feasibility of `Thread.kill`

Reply:
This is an interesting read on killing threads: https://docs.oracle.com/javase/1.5.0/docs/guide/misc/threadPrimitiveDeprecation.html

Reply:
Although, it seems that people are not scared of mentioning `Thread.kill`: https://stackoverflow.com/questions/35001959/interrupt-a-call-in-ocaml

Reply:
But I trust more Janestreet in general, and they don’t like it: https://github.com/janestreet/jbuilder-universe/blob/8e18d43cd478f832967c52cf179e3b72659d97f9/packages/core.v0.9.1/src/core_thread.mli